proof/refine/Move_R.thy

```diff
@@ -457,15 +457,9 @@ lemma bind_dummy_ret_val:
    od = do a; b od"
   by simp
 
-(* FIXME RT: This should replace lookup_cap_and_slot_valid_fault2 *)
-lemma lookup_cap_and_slot_valid_fault3[wp]:
-  "\<lbrace>invs and K (p = to_bl p')\<rbrace>
-   lookup_cap_and_slot thread p
-   -,\<lbrace>\<lambda>ft s. valid_fault (ExceptionTypes_A.CapFault p' rp ft)\<rbrace>"
-  using lookup_cap_and_slot_valid_fault
-  apply (clarsimp simp: validE_E_def validE_def valid_def
-                 split: sum.splits)
-  apply (drule invs_valid_objs, fastforce)
-  done
+(* An "excluded middle" lemma for sc_at_ppred *)
+lemma sc_at_ppred_exm:
+  "sc_at g s \<Longrightarrow> sc_at_ppred p P g s = (\<not> sc_at_ppred p (\<lambda>x. \<not> P x) g s)"
```

Comment:
I am not sure how this lemma is actually used in this PR, but it might be worth noting that the `sc_at` assumption is not needed for the implication from LH to RH. So, if the rewrite is actually used for simplification in this direction, we could do without this assumption and just have an imp lemma.

We do have lemmas similar to this one, such as:
```
lemma not_obj_at'_strengthen:
  "obj_at' (Not ∘ P) p s ⟹ ¬ obj_at' P p s"

lemma not_pred_tcb_at'_strengthen:
  "pred_tcb_at' f (Not ∘ P) p s ⟹ ¬ pred_tcb_at' f P p s"
```
we could write one for `sc_at_pred` along this line, too, maybe?

Reply:
but I am not too sure about aligning all the negation.

Reply:
I agree that this approach will work if we are using this rewrite in the hypotheses of a proof, for in that case the full strength of the equality is not needed. However, if we are using this rewrite in the goal of a proof, then we would instead need "RH implies LH" which is not generally true. The use cases in my proofs are both substitutions in the goal and so the assumption is needed.

I could write two lemmas:
* LH implies RH, no assumption
* RH implies LH, with assumption
but I'm not sure that would be any neater.

Reply:
If it's ok, I'd like to leave this as it is.

Reply:
yes, I said the assumption is redundant for going LH to RH. This lemma as an equality as it is now is not very good in that the use of it for simplifying LH -> RH will require us to prove an unnecessary condition, won’t it?

I would think having LH -> RH as an implication lemma is good. I am not sure about the other direction. There might be a better formulation. But, if it has to be an equality, I’d think it should be in the form of A = (B & C) (or something like that) rather than B => (A = C).

Reply:
I can do this
```
sc_at_ppred p P g s = (obj_at (λko. ∃sc n. ko = SchedContext sc n) g s ∧ ¬ sc_at_ppred p (λx. ¬ P x) g s)
```
but `sc_at` has something funky about valid sizes which is annoying.

Reply:
Ah, I see. I think it can be `bound (scs_of s g)` in this case, instead of `sc_at`... we need to do some cleaning around these. Can we also use, say, `scp` instead of `g`?

Reply:
but your `obj_at` version is equally good, I think.

Reply:
That should be equally easy for my proofs, is that the kind of thing you had in mind?

Reply:
Cool, thanks.